addOns/client/src/main/java/org/zaproxy/addon/client/internal/ClientMapWriter.java

```diff
@@ -0,0 +1,202 @@
+        }
+        if (!node.getUserObject().isVisited()) {
+            outputKV(fw, indent, false, VISITED_KEY, node.getUserObject().isVisited());
+        }
+        if (node.getUserObject().getComponents() != null
+                && !node.getUserObject().getComponents().isEmpty()) {
+            for (ClientSideComponent component : node.getUserObject().getComponents()) {
+                if (component.getTypeForDisplay().equalsIgnoreCase(ClientSideComponent.REDIRECT)) {
+                    outputKV(
+                            fw,
+                            indent,
+                            false,
+                            ClientSideComponent.REDIRECT.toLowerCase(Locale.ROOT),
+                            component.getHref());
+                }
+                if (component
+                        .getTypeForDisplay()
+                        .equalsIgnoreCase(ClientSideComponent.CONTENT_LOADED)) {
+                    outputKV(fw, indent, false, "contentLoaded", true);
+                }
+            }
+            indent = outputComponents(fw, node.getUserObject().getComponents(), level, indent);
+        }
+
+        Stats.incCounter(PopupMenuExportClientMap.STATS_EXPORT_CLIENTMAP + ".node");
+
+        if (node.getChildCount() > 0) {
+            fw.write(indent);
+            fw.write("  ");
+            fw.write(CHILDREN_KEY);
+            fw.write(":");
+            fw.write('\n');
+            node.children()
+                    .asIterator()
+                    .forEachRemaining(
+                            c -> {
+                                try {
+                                    outputNode(fw, (ClientNode) c, level + 1);
+                                } catch (IOException e) {
+                                    LOGGER.error(e.getMessage(), e);
+                                }
+                            });
+        }
+    }
+
+    private static String outputComponents(
+            Writer fw, Set<ClientSideComponent> components, int level, String indent)
+            throws IOException {
+        fw.write(indent + "  ");
+        fw.write("components:\n");
+        indent = " ".repeat((level + 1) * 2);
+        boolean first = true;
+
+        LinkedHashSet<ClientSideComponent> sortedComponents =
+                components.stream().sorted().collect(Collectors.toCollection(LinkedHashSet::new));
+        for (ClientSideComponent component : sortedComponents) {
+            if ((component.getTypeForDisplay() != null
+                            && COMPONENT_TYPES_TO_SKIP_LC.contains(
+                                    component.getTypeForDisplay().toLowerCase(Locale.ROOT))
+                    || (component.getType() != null
+                            && "nodeAdded".equalsIgnoreCase(component.getType())))) {
+                continue;
+            }
+            first = outputKV(fw, indent, first, "nodeType", component.getTypeForDisplay());
```

Comment:
We shouldn't be using i18n'ed strings in the export.

Reply:
I understand your point and I'm happy to change it somehow, this is what was asked for.

Shall I switch display types to an enum so the export can use enum names while the UI uses i18n values?

Reply:
That would be better but it depends how much time that adds.

Reply:
I'm open to suggestions.

Reply:
If there are no time constraints anymore +1 to enum.

Reply:
I'm curious why you feel an enum will be slow?

Reply:
Or do you mean implementation time?

Reply:
The latter.